src/backend/distributed/executor/multi_utility.c

```diff
@@ -3760,3 +3982,90 @@ ProcessDropTableStmt(DropStmt *dropTableStatement)
 		}
 	}
 }
+
+
+/*
+ * ShouldExecuteAlterTableSequentially checks if the given ALTER TABLE
+ * statements should be executed sequentially when there is a foreign
+ * constraint from a distributed table to a reference table.
+ * In case of a column related ALTER TABLE operation, we check explicitly
+ * if there is a foreign constraint on this column from/to a reference table.
+ * Additionally, if the command is run inside a transaction block, we call
+ * SetLocalMultiShardModifyModeToSequential so that the further commands
+ * in the same transaction uses the same connections and does not error out.
+ */
+static bool
+ShouldExecuteAlterTableSequentially(Oid relationId, AlterTableCmd *command)
+{
+	bool executeSequentially = false;
+	AlterTableType alterTableType = command->subtype;
+	if (alterTableType == AT_DropConstraint)
+	{
+		char *constraintName = command->name;
+		if (ConstraintIsAForeignKeyToReferenceTable(constraintName, relationId))
+		{
+			executeSequentially = true;
+		}
+	}
+	else if (alterTableType == AT_DropColumn || alterTableType == AT_AlterColumnType)
+	{
+		char *affectedColumnName = command->name;
+
+		if (ColumnAppearsInForeignKeyToReferenceTable(affectedColumnName,
+													  relationId))
+		{
+			if (IsTransactionBlock() && alterTableType == AT_AlterColumnType)
+			{
+				SetLocalMultiShardModifyModeToSequential();
```

Comment:
Yes, this function seems a bit tricky to me as well. I'll try to simplify it or at least comment better.

`ALTER COLUMN` is tricky because any operation on the columns involving the foreign key runs a `SELECT` on the referencing table. 

So, if we do not switch to the sequential mode in a transaction block, we'd end up with the restrictions where `Parallel SELECT on hash` followed by `DDL` on a reference table, which we cannot support. So, by switching to sequential mode, we ensure that the `SELECT` happens on a single connection per node.

And, note that we don't have to call `SetLocalMultiShardModifyModeToSequential();` for all checks here. For example, `ALTER TABLE ADD FKEY` command itself should be executed sequentially, but, we don't really have to set the multishard mode to sequential which could effect the unrelated consecutive operations in the same transaction. Any dangerous operation is in the transaction is already going to be detected and acted accordingly with the `Enforce Restrictions` PR